DB/Scripts: Terokkar Forest SAI migration

mysterious-static opened the pull request:
**Changes proposed:**

-  Migrate 5 scripts to SAI

**Target branch(es):** 3.3.5/master

- [X] 3.3.5
- [X] master

**Tests performed:** (Does it build, tested in-game, etc.)
- SQL is syntactically valid
- Scripts all tested in game

**TODOS**
- [x] Slim's conditions are technically unnecessary because many vendors skip their gossip menu when conditions are met now. I think it makes sense to have them there (like #20543 / ab8192e) in case someone messes with those npc flags but that's a dev decision.
- [x] SQL consolidation: see 2cf90cd for uncondensed script.

Commit:
First pass

Commit:
Updated list of supported quests

Commit:
Update what can be removed after migration

Commit:
rename and consolidation

Commit:
p sure this needs castflags 2 since the npc is dead

Commit:
floon, and remove some unused defines

Commit:
clean up terokkar forest cpp

Commit:
an unforgivable oversight

Commit:
comments for possible migration for unkor

Commit:
give up on Unkor

One of the conditions for the quest completion isn't available in
DB scripting that I can see. Everything else would work, but
unfortunately I think we have to give this one up

Commit:
smartai for floon

Commit:
pointless comment

Commit:
whoops heh

floon needed to have smartscripts renumbered, i copy/pasted from another sai but didnt fix them

Commit:
sql

Commit:
all set - add todos

Commit:
whoops forgot to save

Aokromes commented on src/server/game/Miscellaneous/SharedDefines.h:
```diff
@@ -169,7 +169,6 @@ enum FactionTemplates
     FACTION_ORC_DRAGONMAW               = 62,
     FACTION_HORDE_GENERIC               = 83,
     FACTION_ALLIANCE_GENERIC            = 84,
-    FACTION_DEMON                       = 90,
```
no need of this changes (this and next change on this file)

velinath replied:
Ok. I took them out because they are no longer used in source. Better to leave them in?

tkrokli replied:
The general rule is to limit our changes to the related code.
Other changes creates "noise" and makes review harder.

velinath replied:
So noted - maybe I'll clean up shareddefines later. Will revert this shortly

offl commented on sql/updates/world/3.3.5/9999_99_99_99_world.sql:
```diff
@@ -0,0 +1,85 @@
+
+DELETE FROM `smart_scripts` WHERE (source_type = 0 AND entryorguid = 22419);
+INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `t ...
+(22419, 0, 0, 0, 54, 0, 100, 0, 0, 0, 0, 0, 49, 0, 0, 0, 0, 0, 0, 21, 15, 0, 0, 0, 0, 0, 0, 'Wood Mite - On Just Summoned - Start Attacking');
```
They shouldn't

velinath replied:
Thank you! I will remove this then.

tkrokli commented:
```cpp
/home/travis/build/TrinityCore/TrinityCore/src/server/scripts/Outland/zone_terokkar_forest.cpp:259:29:
 fatal error: member access into incomplete type 'GameObject'
                        Cage->SetGoState(GO_STATE_ACTIVE);
                            ^
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Maps/ZoneScript.h:25:7: note: forward declaration of 'GameObject'
class GameObject;
      ^
1 error generated.
```
This is in the script `npc_isla_starmane` outside of your changes. Maybe your branch is outdated?

velinath commented:
Nope, I'm just an idiot

offl commented on sql/updates/world/3.3.5/9999_99_99_99_world.sql:
```diff
@@ -0,0 +1,85 @@
+(185913, 1, 4, 5, 62, 0, 100, 0, 8660, 2, 0, 0, 72, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Skull Pile - On Gossip Option 2 Selected - Close Gossip'),
+(185913, 1, 5, 0, 61, 0, 100, 0, 0, 0, 0, 0, 85, 40642, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Skull Pile - On Gossip Option 2 Selected - Invoker Cast \'Summon Darkscreecher Akkarai\''),
+(185913, 1, 6, 7, 62, 0, 100, 0, 8660, 3, 0, 0, 72, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Skull Pile - On Gossip Option 3 Selected - Close Gossip'),
+(185913, 1, 7, 0, 61, 0, 100, 0, 0, 0, 0, 0, 85, 40644, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Skull Pile - On Gossip Option 3 Selected - Invoker Cast \'Summon Vakkiz the Windrager\'');
```
> Gossip does not close properly on Skull Pile (GO 185913) despite event linking working fine. Would appreciate dev eyes, I can't figure it out

Try to close gossip only after spellcast. Actually it's standart for such scripts and maybe will fix the issue.

velinath replied:
this didn't seem to fix it, it just closed the gossip w/o invoker casting. there are no conditions in my DB that would cause this so I'm at a loss

velinath replied:
fwiw, I tried

```sql
UPDATE `gameobject_template` SET `ScriptName` = '', `AIName` = 'SmartGameObjectAI' WHERE `entry` = 185913;
DELETE FROM `smart_scripts` WHERE `entryorguid` = 185913 AND `source_type` = 1;
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(185913, 1, 0, 4, 62, 0, 100, 0, 8660, 0, 0, 0, 85, 40632, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Skull Pile - On Gossip Option 0 Selected - Invoker Cast \'Summon Gezzarak the Huntress\''),
(185913, 1, 1, 4, 62, 0, 100, 0, 8660, 1, 0, 0, 85, 40640, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Skull Pile - On Gossip Option 1 Selected - Invoker Cast \'Summon Karrog\''),
(185913, 1, 2, 4, 62, 0, 100, 0, 8660, 2, 0, 0, 85, 40642, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Skull Pile - On Gossip Option 2 Selected - Invoker Cast \'Summon Darkscreecher Akkarai\''),
(185913, 1, 3, 4, 62, 0, 100, 0, 8660, 3, 0, 0, 85, 40644, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Skull Pile - On Gossip Option 3 Selected - Invoker Cast \'Summon Vakkiz the Windrager\''),
(185913, 1, 4, 0, 61, 0, 100, 0, 0, 0, 0, 0, 72, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Skull Pile - On Gossip Option 0 Selected - Close Gossip');

tkrokli replied:
OK, so that means to swap the order of the 2 rows above, I guess?

velinath replied:
well, what's in the PR currently is "close gossip, then cast spell". gossip does not close, but spell casts.

the above script reverses that order in general, so "cast spell, then close gossip". Now gossip closes but spell does not cast.

offl replied:
Strange, there's a lot of similar scripts. Can't believe they don't work.

offl replied:
Maybe event_param2 (Cast Flags) aren't alright

velinath replied:
mm maybe? but if you reverse the order (have the cast as the linked event) it casts properly

velinath replied:
actually yeah castflags 4 fixes it, but I dont know if thats acceptable

velinath replied:
I'm having trouble reproing it not working now, actually.

tkrokli replied:
I remember some of my SAI scripts (a year back ++) often had a tendency to not work for me, but work OK for others (at least during writing and testing).

velinath replied:
`.reload smart_scripts` didn't work properly. I had to unload the grid and it worked.

Commit:
rm wood mite per feedback

thanks offl

Commit:
rv changes made here

per Aokromes

Commit:
see if fixes travis

if not i'll add the other header back in

Commit:
fixes issue w smart scripting

Commit:
remove todo

There is SmartAI for the summoned mobs so I am assuming they are scripted properly. Probably no need to check this after all.

Commit:
condense sql

Commit:
parens

Killyana commented on sql/updates/world/3.3.5/9999_99_99_99_world.sql:
```diff
@@ -0,0 +1,49 @@
+DELETE FROM `smart_scripts` WHERE (`source_type` = 0 AND `entryorguid` IN (22095, 22307, 18588)) OR (`source_type` = 1 AND `entryorguid` = 185913);
+INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `t ...
+(22095, 0, 0, 0, 6, 0, 75, 0, 0, 0, 0, 0, 11, 39130, 2, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 'Infested Root-Walker - On Just Died - Cast \'Summon Wood Mites\''),
+(22307, 0, 0, 0, 6, 0, 75, 0, 0, 0, 0, 0, 11, 39134, 2, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 'Rotting Forest-Rager - On Just Died - Cast \'Summon Lots Of Wood Mites\''),
```
Add this combat spell http://www.wowhead.com/spell=15548/thunderclap

Killyana commented on sql/updates/world/3.3.5/9999_99_99_99_world.sql:
```diff
@@ -0,0 +1,49 @@
+
+DELETE FROM `smart_scripts` WHERE (`source_type` = 0 AND `entryorguid` IN (22095, 22307, 18588)) OR (`source_type` = 1 AND `entryorguid` = 185913);
+INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `t ...
+(22095, 0, 0, 0, 6, 0, 75, 0, 0, 0, 0, 0, 11, 39130, 2, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 'Infested Root-Walker - On Just Died - Cast \'Summon Wood Mites\''),
```
Add this combat spell http://www.wowhead.com/spell=39000/regrowth

tkrokli replied:
Spell is added to the line below these comments
(hence these comments are not hidden).

Killyana commented:
Keep the condition for Slim, just add the 2 missing combat  spells every thing else is fine.

velinath commented:
Do you have something showing when those spells should be cast?

Killyana commented:
Use this:
``` sql
 2209501, 22095, 0, 0, 100, 1, 15000, 20000, 32000, 38000, 11, 39000, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, Infested Root-Walker - CAST Regrowth
 2230701, 22307, 9, 0, 100, 1, 0, 5, 12000, 18000, 11, 15548, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, Rotting Forest-Rager - CAST Thunderclap

velinath commented:
perfect - thanks

Commit:
add casts based on noted events

per @Killyana, just threw it into update IC, obviously remove the norepeat flag if i'm being dumb but since it was in the original supplied stuff i figured it should be here too

if there's time i'll poke it tonight to make sure but otherwise this should be _fine_

Commit:
sai id needs to be unique, whoops

tkrokli commented on sql/updates/world/3.3.5/9999_99_99_99_world.sql:
```diff
@@ -0,0 +1,51 @@
+DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 15 AND `SourceGroup` IN (7732, 8660);
+DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` IN (14, 15) AND `SourceGroup` = 8021;
+DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 22 AND `SourceGroup` = 1 AND `SourceEntry` IN (22095, 22307) AND `SourceId` = 0;
+
```
Cosmetic: this blank line can be removed to group the related queries.

Killyana committed:
Update 9999_99_99_99_world.sql

Aokromes committed:
Rename 9999_99_99_99_world.sql to 2017_11_20_01_world.sql
